ui/src/criteria/attribute.tsx

```diff
@@ -81,23 +133,145 @@ function AttributeEdit(props: AttributeEditProps) {
   const hintDisplayName = (hint: tanagra.EnumHintValue) =>
     hint.displayName || "Unknown Value";
 
-  const enumHintValues = underlay.entities
+  const attributeFilterHint = underlay.entities
     .find((g) => g.name === underlay.primaryEntity)
-    ?.attributes?.find((attribute) => attribute.name === props.data.attribute)
-    ?.attributeFilterHint?.enumHint?.enumHintValues;
+    ?.attributes?.find(
+      (attribute) => attribute.name === props.data.attribute
+    )?.attributeFilterHint;
+  const enumHintValues = attributeFilterHint?.enumHint?.enumHintValues;
+  const integerBoundsHint = attributeFilterHint?.integerBoundsHint;
 
-  const selectionIndex = (hint: tanagra.EnumHintValue) =>
-    props.data.selected.findIndex(
-      (row) => row.id === hint.attributeValue?.int64Val
-    );
+  const hasValidIntegerBounds =
+    isValid(integerBoundsHint?.min) && isValid(integerBoundsHint?.max);
 
-  if (enumHintValues?.length === 0) {
+  if (!hasValidIntegerBounds && enumHintValues?.length === 0) {
     return (
       <Typography>
         No information for attribute {props.data.attribute}.
       </Typography>
     );
   }
+
+  if (hasValidIntegerBounds) {
+    const minBound = integerBoundsHint?.min || 0; // This is to ensure the compiler won't complain the object might be undefined
+    const maxBound = integerBoundsHint?.max || 0; // although we already know that min and max is valid
+
+    //two sets of values are needed due to the input box and slider is isolated.
+    const [minInputValue, setMinInputValue] = useState<number>(minBound);
+    const [maxInputValue, setMaxInputValue] = useState<number>(maxBound);
+    const [minValue, setMinValue] = useState<number>(minBound);
+    const [maxValue, setMaxValue] = useState<number>(maxBound);
+
+    const updateValue = (newMin: number, newMax: number) => {
+      setMinValue(newMin);
+      setMaxValue(newMax);
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleLeftInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin =
+        event.target.value === "" ? minBound : Number(event.target.value);
+      setMinInputValue(newMin);
+
+      newMin = Math.max(minBound, newMin);
+      newMin = Math.min(maxValue, newMin);
+      updateValue(newMin, maxValue);
+    };
+    const handleLeftInputBlur = () => {
```

Comment:
https://mui.com/material-ui/api/input-base/#props this is the definition for onBlur. it like an opposite to onFocus. 
it will be called when it unfocus. 
(I know it is tricky, but in order to solve the issue comparing while type "4"  to "40". it will need a kind of stage like onblur and ensure the number input is finished)

Reply:
I see. The `InputBase` docs weren't linked from the slider docs so I didn't see them. It makes sense to handle that case though.